

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Analysing lung cells from aging mice with scllm &#8212; scllm 0.2.0 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css" />
    <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'notebooks/angelidis';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Analysing Xenium data with scllm" href="xenium.html" />
    <link rel="prev" title="Identifying gene signatures in Visium data using NMF and scllm" href="nmf.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">scllm 0.2.0 documentation</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Tutorials:</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="interface.html">Quick Walkthrough</a></li>
<li class="toctree-l1"><a class="reference internal" href="pbmc3k.html">Annotating the PMBC3K dataset using PCA and scllm</a></li>
<li class="toctree-l1"><a class="reference internal" href="nmf.html">Identifying gene signatures in Visium data using NMF and scllm</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Analysing lung cells from aging mice with scllm</a></li>
<li class="toctree-l1"><a class="reference internal" href="xenium.html">Analysing Xenium data with scllm</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">API Reference:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../api.html">API Reference</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../_autosummary/scllm.tl.html">scllm.tl</a></li>
<li class="toctree-l2"><a class="reference internal" href="../_autosummary/scllm.pl.html">scllm.pl</a></li>
</ul>
</details></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/sagar87/scllm" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/sagar87/scllm/edit/main/docs/notebooks/angelidis.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/sagar87/scllm/issues/new?title=Issue%20on%20page%20%2Fnotebooks/angelidis.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/notebooks/angelidis.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Analysing lung cells from aging mice with scllm</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Summary">Summary</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Setup">Setup</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Data-preprocessing">Data preprocessing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Annotating-cell-types">Annotating cell types</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="Analysing-lung-cells-from-aging-mice-with-scllm">
<h1>Analysing lung cells from aging mice with scllm<a class="headerlink" href="#Analysing-lung-cells-from-aging-mice-with-scllm" title="Permalink to this heading">#</a></h1>
<section id="Summary">
<h2>Summary<a class="headerlink" href="#Summary" title="Permalink to this heading">#</a></h2>
<p>In this code, we analyze single-cell RNA sequencing data using the <code class="docutils literal notranslate"><span class="pre">scanpy</span></code> library and a language model for annotation. We start by loading our dataset and then use a pre-trained language model to annotate factors based on principal component analysis (PCA) results. We visualize these factors with strip plots and embeddings. Next, we compute neighborhood relationships and identify clusters using the Leiden algorithm. We further annotate these clusters with cell type information using the
language model and visualize the results through embeddings and heatmaps. This workflow allows us to gain insights into the cellular composition and characteristics of our dataset.</p>
</section>
<section id="Setup">
<h2>Setup<a class="headerlink" href="#Setup" title="Permalink to this heading">#</a></h2>
<p>In this section, we set up our environment to work with the <code class="docutils literal notranslate"><span class="pre">scllm</span></code> library and the Angelidis et al. dataset. First, we load our API key for ChatGPT using <code class="docutils literal notranslate"><span class="pre">dotenv</span></code>, which allows us to keep our credentials secure and separate from our code. Then, we import necessary libraries: <code class="docutils literal notranslate"><span class="pre">scllm</span></code> for single-cell analysis, <code class="docutils literal notranslate"><span class="pre">scanpy</span></code> for handling our scRNA-seq data, <code class="docutils literal notranslate"><span class="pre">pandas</span></code> for data manipulation, and <code class="docutils literal notranslate"><span class="pre">matplotlib</span></code> for plotting.</p>
<p>Next, we read the Angelidis dataset from an HDF5 file using <code class="docutils literal notranslate"><span class="pre">sc.read_h5ad()</span></code>, which loads our single-cell data into an AnnData object. This sets the stage for further analysis, such as annotating factors and clusters in the dataset.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Enter your API key for ChatGPT</span>
<span class="c1"># import os</span>
<span class="c1"># os.environ[&quot;OPENAI_API_KEY&quot;] = &quot;Enter your API key here.&quot;</span>
<span class="c1"># Or alternatively save your OPENAI key in a .env file and load it via dotenv</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">dotenv</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_dotenv</span>
<span class="n">load_dotenv</span><span class="p">(</span><span class="s2">&quot;../../../../.env&quot;</span><span class="p">)</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scllm</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sl</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">textwrap</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">langchain_openai</span><span class="w"> </span><span class="kn">import</span> <span class="n">ChatOpenAI</span>
</pre></div>
</div>
</div>
<p>In these lines, we set up the OpenAI model that we’ll use for our analysis. The variable <code class="docutils literal notranslate"><span class="pre">openai_model</span></code> is assigned the string <code class="docutils literal notranslate"><span class="pre">&quot;gpt-4o-mini&quot;</span></code>, which specifies the version of the model we want to use. Then, we create an instance of <code class="docutils literal notranslate"><span class="pre">ChatOpenAI</span></code> called <code class="docutils literal notranslate"><span class="pre">llm</span></code>, where we set the <code class="docutils literal notranslate"><span class="pre">temperature</span></code> to <code class="docutils literal notranslate"><span class="pre">0.0</span></code>. This means we want the model to generate consistent and deterministic outputs, rather than random or creative ones. This is crucial for tasks like annotating factors and clusters in our
single-cell RNA-seq dataset, as we want reliable and reproducible results. Essentially, this setup allows us to leverage the power of the OpenAI model for our computational biology tasks.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">openai_model</span> <span class="o">=</span> <span class="s2">&quot;gpt-4o-mini&quot;</span>
<span class="n">llm</span> <span class="o">=</span> <span class="n">ChatOpenAI</span><span class="p">(</span><span class="n">temperature</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="n">openai_model</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Data-preprocessing">
<h2>Data preprocessing<a class="headerlink" href="#Data-preprocessing" title="Permalink to this heading">#</a></h2>
<p>We start by loading the Angelidis dataset, which is a complex single-cell RNA-seq dataset containing over 20 different cell types from mouse lung tissue. This dataset is already preprocessed and includes a precomputed PCA, making our analysis easier.</p>
<p>The line <code class="docutils literal notranslate"><span class="pre">angelidis</span> <span class="pre">=</span> <span class="pre">sc.read_h5ad('/Users/harald/Downloads/angelidis.h5ad')</span></code> reads the dataset from an H5AD file into an AnnData object, which is a convenient format for handling single-cell data in Python. This object allows us to access the expression data, metadata, and precomputed PCA results efficiently. With this setup, we can proceed to annotate factors and clusters, leveraging the power of the <code class="docutils literal notranslate"><span class="pre">scllm</span></code> library and the ChatGPT model for insightful biological interpretations.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">angelidis</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">&#39;/Users/harald/Downloads/angelidis.h5ad&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>In this snippet, we use <code class="docutils literal notranslate"><span class="pre">sl.tl.annotate_factor</span></code> to identify cell types based on the first principal components (PCs) from our scRNA-seq data. By default, this function queries for cell types, which is crucial since the first components often represent distinct cell types. We set <code class="docutils literal notranslate"><span class="pre">top_genes=50</span></code> to focus on the top 50 genes contributing to each factor, and <code class="docutils literal notranslate"><span class="pre">num_samples=10</span></code> tells the function to query these genes 10 times. Scllm then extracts the most likely term for each factor, giving us a
clearer understanding of the underlying biology. This step is essential for annotating our dataset and helps us interpret the cellular diversity in the lung tissue from the Angelidis et al. study.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">annotate_factor</span><span class="p">(</span>
    <span class="n">angelidis</span><span class="p">,</span>
    <span class="n">varm_key</span><span class="o">=</span><span class="s1">&#39;PCs&#39;</span><span class="p">,</span>
    <span class="n">llm</span><span class="o">=</span><span class="n">llm</span><span class="p">,</span>
    <span class="n">top_genes</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
    <span class="n">num_samples</span><span class="o">=</span><span class="mi">10</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>In this part of the script, we create a factor strip plot using <code class="docutils literal notranslate"><span class="pre">sl.pl.factor_stripplot</span></code>. This visualization helps us get an overview of the major axes of variation in our single-cell RNA-seq dataset. The factors we’re plotting (0 to 5) are likely driven by differences across cell types, which is crucial for understanding the biological context of our data.</p>
<p>We specify the <code class="docutils literal notranslate"><span class="pre">obsm_key='X_pca'</span></code> to use the PCA coordinates for our plot, and <code class="docutils literal notranslate"><span class="pre">annotation_key='factor_annotation'</span></code> to label the points based on their factor annotations. The <code class="docutils literal notranslate"><span class="pre">size</span></code> and <code class="docutils literal notranslate"><span class="pre">jitter</span></code> parameters help us adjust the appearance of the points, making it easier to see the distribution. By plotting these factors, we can quickly identify how different cell types contribute to the overall variation in the dataset.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
<span class="n">sl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">factor_stripplot</span><span class="p">(</span>
    <span class="n">angelidis</span><span class="p">,</span>
    <span class="n">factors</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;0&#39;</span><span class="p">,</span> <span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="s1">&#39;2&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">,</span> <span class="s1">&#39;4&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">],</span>
    <span class="n">obsm_key</span><span class="o">=</span><span class="s1">&#39;X_pca&#39;</span><span class="p">,</span>
    <span class="n">annotation_key</span><span class="o">=</span><span class="s1">&#39;factor_annotation&#39;</span><span class="p">,</span>
    <span class="n">size</span><span class="o">=</span><span class="mf">.4</span><span class="p">,</span>
    <span class="n">jitter</span><span class="o">=</span><span class="mf">.3</span><span class="p">,</span>
    <span class="n">ax</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_angelidis_11_0.png" src="../_images/notebooks_angelidis_11_0.png" />
</div>
</div>
<p>In this snippet, we use <code class="docutils literal notranslate"><span class="pre">sl.pl.factor_embedding</span></code> to overlay UMAP plots with factor weights derived from our PCA analysis. By specifying <code class="docutils literal notranslate"><span class="pre">obsm_key='X_pca'</span></code>, we’re telling the function to use the PCA coordinates for our cells. The <code class="docutils literal notranslate"><span class="pre">factors</span></code> parameter includes the top factors we identified earlier, which represent important features in our dataset.</p>
<p>Overlaying these factors on the UMAP allows us to visualize how different cell populations relate to these key features, helping us understand the underlying biology better. The <code class="docutils literal notranslate"><span class="pre">annotation_key='factor_annotation'</span></code> links our factors to meaningful biological interpretations, making it easier to see how they influence cell distribution in the UMAP space. The <code class="docutils literal notranslate"><span class="pre">ncols=2</span></code> just organizes our plots into two columns for better readability.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">sl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">factor_embedding</span><span class="p">(</span>
    <span class="n">angelidis</span><span class="p">,</span>
    <span class="n">obsm_key</span><span class="o">=</span><span class="s1">&#39;X_pca&#39;</span><span class="p">,</span>
    <span class="n">factors</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;0&#39;</span><span class="p">,</span> <span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="s1">&#39;2&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">,</span> <span class="s1">&#39;4&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">],</span>
    <span class="n">basis</span><span class="o">=</span><span class="s1">&#39;X_pca_umap&#39;</span><span class="p">,</span>
    <span class="n">annotation_key</span><span class="o">=</span><span class="s1">&#39;factor_annotation&#39;</span><span class="p">,</span>
    <span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_angelidis_13_0.png" src="../_images/notebooks_angelidis_13_0.png" />
</div>
</div>
</section>
<section id="Annotating-cell-types">
<h2>Annotating cell types<a class="headerlink" href="#Annotating-cell-types" title="Permalink to this heading">#</a></h2>
<p>In this part of the script, we first call <code class="docutils literal notranslate"><span class="pre">sc.pp.neighbors(angelidis)</span></code>, which computes the neighborhood graph of the cells based on their PCA embeddings. This step is crucial because it helps us understand how cells are related to each other in the high-dimensional space. Next, we use <code class="docutils literal notranslate"><span class="pre">sc.tl.leiden(angelidis)</span></code> to perform community detection using the Leiden algorithm, which identifies clusters of similar cells. These clusters represent different cell types or states within our dataset. By
doing this, we set the stage for further analysis and annotation of these clusters, allowing us to assign meaningful biological identities to the cell populations we’ve identified.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">angelidis</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span><span class="n">angelidis</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/var/folders/_4/qhxlw4x53134ffft2v5pt3nh0000gn/T/ipykernel_58649/144542317.py:2: FutureWarning: In the future, the default backend for leiden will be igraph instead of leidenalg.

 To achieve the future defaults please pass: flavor=&#34;igraph&#34; and n_iterations=2.  directed must also be False to work with igraph&#39;s implementation.
  sc.tl.leiden(angelidis)
</pre></div></div>
</div>
<p>In this part of the script, we want to inspect the clusters we’ve identified in our single-cell RNA-seq data. The line <code class="docutils literal notranslate"><span class="pre">sc.pl.embedding(angelidis,</span> <span class="pre">basis=&quot;X_pca_umap&quot;,</span> <span class="pre">color='leiden')</span></code> visualizes these clusters using the UMAP representation of our PCA results, coloring the points by their cluster assignments from the Leiden algorithm.</p>
<p>Next, we aim to annotate the cell types within these clusters. The line <code class="docutils literal notranslate"><span class="pre">sl.tl.annotate_cluster(angelidis,</span> <span class="pre">cluster_key='leiden',</span> <span class="pre">llm=llm,</span> <span class="pre">num_samples=5,</span> <span class="pre">key_added=&quot;cell_types_1&quot;)</span></code> uses the <code class="docutils literal notranslate"><span class="pre">scllm</span></code> library to automatically assign cell type annotations to each cluster based on the Leiden labels. We specify that we want 5 samples for the annotation, and the results will be stored under the key “cell_types_1”. This helps us understand the biological significance of the clusters we’ve visualized.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">angelidis</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;X_pca_umap&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_angelidis_17_0.png" src="../_images/notebooks_angelidis_17_0.png" />
</div>
</div>
<p>In this part of the script, we use the <code class="docutils literal notranslate"><span class="pre">sl.tl.annotate_cluster</span></code> function to assign cell type annotations to the clusters identified in our scRNA-seq dataset. We specify the <code class="docutils literal notranslate"><span class="pre">cluster_key</span></code> as ‘leiden’, which refers to the clustering results we obtained earlier. The <code class="docutils literal notranslate"><span class="pre">llm</span></code> parameter is our language model, which helps generate meaningful annotations based on the cluster characteristics. We set <code class="docutils literal notranslate"><span class="pre">num_samples</span></code> to 5, meaning we want five different annotations for each cluster, and we store the
results in a new key called “cell_types_1”. This step is crucial as it helps us interpret the biological significance of the clusters, giving us insights into the different cell types present in the lung tissue from the Angelidis dataset.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">annotate_cluster</span><span class="p">(</span>
    <span class="n">angelidis</span><span class="p">,</span>
    <span class="n">cluster_key</span><span class="o">=</span><span class="s1">&#39;leiden&#39;</span><span class="p">,</span>
    <span class="n">llm</span><span class="o">=</span><span class="n">llm</span><span class="p">,</span>
    <span class="n">num_samples</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
    <span class="n">key_added</span><span class="o">=</span><span class="s2">&quot;cell_types_1&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/harald/Documents/opt/scllm/.venv/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, &#34;names&#34;] = self.var_names[global_indices]
/Users/harald/Documents/opt/scllm/.venv/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, &#34;scores&#34;] = scores[global_indices]
/Users/harald/Documents/opt/scllm/.venv/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, &#34;pvals&#34;] = pvals[global_indices]
/Users/harald/Documents/opt/scllm/.venv/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, &#34;pvals_adj&#34;] = pvals_adj[global_indices]
/Users/harald/Documents/opt/scllm/.venv/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, &#34;logfoldchanges&#34;] = np.log2(
/Users/harald/Documents/opt/scllm/.venv/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, &#34;names&#34;] = self.var_names[global_indices]
/Users/harald/Documents/opt/scllm/.venv/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, &#34;scores&#34;] = scores[global_indices]
/Users/harald/Documents/opt/scllm/.venv/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, &#34;pvals&#34;] = pvals[global_indices]
/Users/harald/Documents/opt/scllm/.venv/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, &#34;pvals_adj&#34;] = pvals_adj[global_indices]
/Users/harald/Documents/opt/scllm/.venv/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, &#34;logfoldchanges&#34;] = np.log2(
/Users/harald/Documents/opt/scllm/src/scllm/tl/cluster_annotation.py:40: FutureWarning: the convert_dtype parameter is deprecated and will be removed in a future version.  Do ``ser.astype(object).apply()`` instead if you want ``convert_dtype=False``.
  flattened=lambda df: df[&#34;union&#34;].apply(
</pre></div></div>
</div>
<p>In this part of the script, we’re visualizing the results of our cell type annotations on the UMAP embedding of the single-cell RNA-seq data. The line <code class="docutils literal notranslate"><span class="pre">sc.pl.embedding(angelidis,</span> <span class="pre">basis=&quot;X_pca_umap&quot;,</span> <span class="pre">color='cell_types_1')</span></code> generates a plot where each point represents a single cell, colored according to the cell types we just annotated using the <code class="docutils literal notranslate"><span class="pre">sl.tl.annotate_cluster</span></code> function. This helps us see how well our clustering aligns with the identified cell types, giving us insights into the
cellular composition of the lung tissue. Essentially, we’re making it easier to interpret our data by visually representing the different cell types in a meaningful way.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">angelidis</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;X_pca_umap&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;cell_types_1&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_angelidis_21_0.png" src="../_images/notebooks_angelidis_21_0.png" />
</div>
</div>
<p>In this part of the code, we’re generating a heatmap to visualize the expression of genes across different cell types identified in our single-cell RNA-seq dataset. We use <code class="docutils literal notranslate"><span class="pre">sc.pl.heatmap()</span></code> from the Scanpy library, where we specify <code class="docutils literal notranslate"><span class="pre">groupby=&quot;cell_types_1&quot;</span></code> to group our data by the cell types we annotated earlier using the <code class="docutils literal notranslate"><span class="pre">scllm</span></code> package. The <code class="docutils literal notranslate"><span class="pre">var_names=angelidis.uns['cell_types_1']['var_names']</span></code> argument pulls the specific genes associated with those cell types from the results stored
in the AnnData object. This heatmap helps us quickly see which genes are upregulated or downregulated in each cell type, giving us insights into their functional roles in the lung tissue.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">angelidis</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;cell_types_1&quot;</span><span class="p">,</span> <span class="n">var_names</span><span class="o">=</span><span class="n">angelidis</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;cell_types_1&#39;</span><span class="p">][</span><span class="s1">&#39;var_names&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: Gene labels are not shown when more than 50 genes are visualized. To show gene labels set `show_gene_labels=True`
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_angelidis_23_1.png" src="../_images/notebooks_angelidis_23_1.png" />
</div>
</div>
<p>In this script, we start by loading the Angelidis dataset, which contains single-cell RNA-seq data from mouse lung tissue. We use <code class="docutils literal notranslate"><span class="pre">scllm</span></code> to annotate factors based on PCA results, pulling in the top 50 genes and generating 10 samples for better insights. Next, we visualize these factors with a strip plot and an embedding plot, helping us see how the cells cluster in PCA space.</p>
<p>Then, we compute the neighborhood graph and apply the Leiden algorithm to identify clusters in the data. After that, we annotate these clusters with cell type information using the LLM, adding this to our dataset. Finally, we visualize the clusters and their corresponding heatmap, giving us a clear view of gene expression patterns across the identified cell types.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="nmf.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Identifying gene signatures in Visium data using NMF and scllm</p>
      </div>
    </a>
    <a class="right-next"
       href="xenium.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Analysing Xenium data with scllm</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Summary">Summary</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Setup">Setup</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Data-preprocessing">Data preprocessing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Annotating-cell-types">Annotating cell types</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Harald Vohringer
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2025, Harald Vohringer.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>