

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Identifying gene signatures in Visium data using NMF and scllm &#8212; scllm 0.1.5 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css" />
    <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'notebooks/nmf';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Analysing lung cells from aging mice with scllm" href="angelidis.html" />
    <link rel="prev" title="Annotating the PMBC3K dataset using PCA and scllm" href="pbmc3k.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">scllm 0.1.5 documentation</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Tutorials:</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="pbmc3k.html">Annotating the PMBC3K dataset using PCA and scllm</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Identifying gene signatures in Visium data using NMF and scllm</a></li>
<li class="toctree-l1"><a class="reference internal" href="angelidis.html">Analysing lung cells from aging mice with scllm</a></li>
<li class="toctree-l1"><a class="reference internal" href="xenium.html">Analysing Xenium data with scllm</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">API Reference:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../api.html">API Reference</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../_autosummary/scllm.tl.html">scllm.tl</a></li>
<li class="toctree-l2"><a class="reference internal" href="../_autosummary/scllm.pl.html">scllm.pl</a></li>
</ul>
</details></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/sagar87/scllm" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/sagar87/scllm/edit/main/docs/notebooks/nmf.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/sagar87/scllm/issues/new?title=Issue%20on%20page%20%2Fnotebooks/nmf.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/notebooks/nmf.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Identifying gene signatures in Visium data using NMF and scllm</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Summary">Summary</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Data-preprocessing">Data preprocessing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Applying-NMF">Applying NMF</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Annotating-factors">Annotating factors</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="Identifying-gene-signatures-in-Visium-data-using-NMF-and-scllm">
<h1>Identifying gene signatures in Visium data using NMF and scllm<a class="headerlink" href="#Identifying-gene-signatures-in-Visium-data-using-NMF-and-scllm" title="Permalink to this heading">#</a></h1>
<section id="Summary">
<h2>Summary<a class="headerlink" href="#Summary" title="Permalink to this heading">#</a></h2>
<p>In this code, we analyze spatial transcriptomics data from a mouse brain sample using various computational biology libraries. We start by loading the necessary libraries and the dataset, which includes gene expression counts and spatial information. We perform quality control by filtering cells and genes based on specific criteria, followed by normalization and identification of highly variable genes. We then apply Non-negative Matrix Factorization (NMF) to reduce the dimensionality of the
data, extracting latent factors that represent gene expression patterns. Finally, we annotate these factors using a language model and visualize the results in spatial plots, allowing us to interpret the biological significance of the identified factors in the context of the tissue architecture.</p>
<p>In this part of the script, we start by loading our environment variables, specifically the OpenAI API key, which is essential for using the ChatGPT model later on. We then import several libraries: <code class="docutils literal notranslate"><span class="pre">scanpy</span></code> for single-cell analysis, <code class="docutils literal notranslate"><span class="pre">scllm</span></code> for factor annotation, <code class="docutils literal notranslate"><span class="pre">squidpy</span></code> for spatial transcriptomics, and <code class="docutils literal notranslate"><span class="pre">seaborn</span></code> and <code class="docutils literal notranslate"><span class="pre">matplotlib</span></code> for visualization.</p>
<p>Next, we set up the ChatGPT model with a specified temperature and model type. This model will help us annotate the factors derived from our NMF analysis. The imports and model setup are crucial for the subsequent steps where we process the 10x Visium mouse brain dataset, perform NMF, and finally annotate the resulting factors using the <code class="docutils literal notranslate"><span class="pre">scllm</span></code> package.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Enter your API key for ChatGPT</span>
<span class="c1"># import os</span>
<span class="c1"># os.environ[&quot;OPENAI_API_KEY&quot;] = &quot;Enter your API key here.&quot;</span>
<span class="c1"># Or alternatively save your OPENAI key in a .env file and load it via dotenv</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">dotenv</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_dotenv</span>
<span class="n">load_dotenv</span><span class="p">()</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">langchain_openai</span><span class="w"> </span><span class="kn">import</span> <span class="n">ChatOpenAI</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scllm</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sl</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">squidpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sq</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.decomposition</span><span class="w"> </span><span class="kn">import</span> <span class="n">NMF</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/harald/Documents/opt/scllm/.venv/lib/python3.12/site-packages/dask/dataframe/__init__.py:31: FutureWarning: The legacy Dask DataFrame implementation is deprecated and will be removed in a future version. Set the configuration option `dataframe.query-planning` to `True` or None to enable the new Dask Dataframe implementation and silence this warning.
  warnings.warn(
/Users/harald/Documents/opt/scllm/.venv/lib/python3.12/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.
  warnings.warn(msg, FutureWarning)
</pre></div></div>
</div>
<p>In these lines, we set up the language model that will help us annotate our data. We specify the model we want to use, which is “gpt-4o-mini,” and then create an instance of <code class="docutils literal notranslate"><span class="pre">ChatOpenAI</span></code> with a temperature of 0.0. This temperature setting means we want the model to generate more deterministic and focused responses, rather than creative or random ones. This is crucial for our task, as we need reliable annotations for the NMF factors derived from our mouse brain dataset. By initializing this
model, we can later call it to interpret the NMF components and provide meaningful biological insights based on the data we processed.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">openai_model</span> <span class="o">=</span> <span class="s2">&quot;gpt-4o-mini&quot;</span>
<span class="n">llm</span> <span class="o">=</span> <span class="n">ChatOpenAI</span><span class="p">(</span><span class="n">temperature</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="n">openai_model</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>In this part of the script, we’re downloading the 10x Genomics mouse brain dataset. The first two <code class="docutils literal notranslate"><span class="pre">wget</span></code> commands fetch the filtered feature barcode matrix and the spatial data files. We then move these files into a <code class="docutils literal notranslate"><span class="pre">data</span></code> directory for organization. After that, we extract the spatial data from the tar.gz file using <code class="docutils literal notranslate"><span class="pre">tar</span> <span class="pre">-xzf</span></code>, which unpacks the contents. Finally, we clean up by removing the tar.gz file since we only need the extracted spatial data. This setup is crucial because it
prepares our data for analysis in the subsequent steps of the script, where we’ll perform quality control, normalization, and NMF analysis.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># %%sh</span>
<span class="c1"># wget https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Brain_Sagittal_Anterior/V1_Mouse_Brain_Sagittal_Anterior_filtered_feature_bc_matrix.h5</span>
<span class="c1"># wget https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Brain_Sagittal_Anterior/V1_Mouse_Brain_Sagittal_Anterior_spatial.tar.gz</span>
<span class="c1"># mv V1_Mouse_Brain_Sagittal_Anterior_filtered_feature_bc_matrix.h5 data</span>
<span class="c1"># mv V1_Mouse_Brain_Sagittal_Anterior_spatial.tar.gz data</span>
<span class="c1"># tar -xzf data/V1_Mouse_Brain_Sagittal_Anterior_spatial.tar.gz</span>
<span class="c1"># rm data/V1_Mouse_Brain_Sagittal_Anterior_spatial.tar.gz</span>
<span class="c1"># mv spatial data</span>
</pre></div>
</div>
</div>
</section>
<section id="Data-preprocessing">
<h2>Data preprocessing<a class="headerlink" href="#Data-preprocessing" title="Permalink to this heading">#</a></h2>
<p>In these lines, we’re loading the 10x Genomics Visium dataset for the mouse brain. The <code class="docutils literal notranslate"><span class="pre">sq.read.visium</span></code> function reads the spatial transcriptomics data from the specified directory, using the provided counts file. This gives us an AnnData object, <code class="docutils literal notranslate"><span class="pre">adata</span></code>, which is a convenient data structure for handling single-cell data.</p>
<p>Next, we call <code class="docutils literal notranslate"><span class="pre">adata.var_names_make_unique()</span></code>, which ensures that all gene names in our dataset are unique. This is crucial because duplicate gene names can cause issues during analysis, especially when we want to reference specific genes later on. So, with these lines, we set the stage for further processing and analysis of our spatial transcriptomics data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">sq</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">visium</span><span class="p">(</span><span class="s2">&quot;data&quot;</span><span class="p">,</span> <span class="n">counts_file</span><span class="o">=</span><span class="s2">&quot;V1_Mouse_Brain_Sagittal_Anterior_filtered_feature_bc_matrix.h5&quot;</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/harald/Documents/opt/scllm/.venv/lib/python3.12/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  utils.warn_names_duplicates(&#34;var&#34;)
/Users/harald/Documents/opt/scllm/.venv/lib/python3.12/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  utils.warn_names_duplicates(&#34;var&#34;)
</pre></div></div>
</div>
<p>In this part of the script, we load a 10x Genomics mouse brain dataset and preprocess it using Squidpy and Scanpy. We start by reading the spatial data and making sure the variable names are unique. Then, we calculate quality control metrics, filtering out low-quality cells and genes. After normalizing and log-transforming the data, we identify highly variable genes. We apply NMF to reduce the dimensionality of the data into 10 components, which we store in the <code class="docutils literal notranslate"><span class="pre">obsm</span></code> and <code class="docutils literal notranslate"><span class="pre">varm</span></code> attributes
of the AnnData object. Finally, we use the <code class="docutils literal notranslate"><span class="pre">scllm</span></code> package to annotate these NMF factors, visualizing the results with strip plots and spatial embeddings to see how the factors relate to the spatial distribution of the data. The <code class="docutils literal notranslate"><span class="pre">adata</span></code> object at the end allows us to inspect the processed data and annotations.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 2695 × 32285
    obs: &#39;in_tissue&#39;, &#39;array_row&#39;, &#39;array_col&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;
    uns: &#39;spatial&#39;
    obsm: &#39;spatial&#39;
</pre></div></div>
</div>
<p>In these lines, we first create a new column in the <code class="docutils literal notranslate"><span class="pre">adata.var</span></code> DataFrame called “mt” that identifies mitochondrial genes. We do this by checking if the gene names start with “mt-”. This is important because mitochondrial gene expression can indicate cell stress or damage.</p>
<p>Next, we calculate quality control (QC) metrics for our dataset using <code class="docutils literal notranslate"><span class="pre">sc.pp.calculate_qc_metrics()</span></code>. We specifically look at the “mt” variable we just created. This function adds several QC metrics to the <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> DataFrame, like the percentage of counts coming from mitochondrial genes, which helps us assess the quality of our single-cell data. By filtering out low-quality cells later, we ensure our analysis is based on reliable data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;mt&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;mt-&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">qc_vars</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;mt&quot;</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>In this part of the code, we create a set of histograms to visualize some key metrics from our spatial transcriptomics data. The first histogram shows the distribution of total counts per cell, giving us an idea of the overall sequencing depth. The second histogram zooms in on cells with total counts below 10,000, helping us identify potential low-quality cells. The third and fourth histograms display the number of genes detected per cell, with the latter focusing on cells with fewer than 4,000
genes. This helps us assess the quality of our data and filter out low-quality cells later on. Finally, we use <code class="docutils literal notranslate"><span class="pre">plt.tight_layout()</span></code> to ensure our plots are neatly arranged without overlapping.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;total_counts&quot;</span><span class="p">],</span> <span class="n">kde</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;total_counts&quot;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;total_counts&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">10000</span><span class="p">],</span> <span class="n">kde</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">40</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">],</span> <span class="n">kde</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">60</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
<span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">4000</span><span class="p">],</span> <span class="n">kde</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">60</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_nmf_16_0.png" src="../_images/notebooks_nmf_16_0.png" />
</div>
</div>
<p>In this part of the script, we’re cleaning up our data to ensure quality before diving into analysis. First, <code class="docutils literal notranslate"><span class="pre">sc.pp.filter_cells(adata,</span> <span class="pre">max_counts=50000)</span></code> filters out cells that have more than 50,000 total counts, which helps us remove potential outliers or low-quality cells. Next, we check for mitochondrial gene expression with <code class="docutils literal notranslate"><span class="pre">adata</span> <span class="pre">=</span> <span class="pre">adata[adata.obs.pct_counts_mt</span> <span class="pre">&lt;</span> <span class="pre">20]</span></code>, keeping only cells where less than 20% of their total counts come from mitochondrial genes. High mitochondrial content
can indicate stressed or dying cells, so we want to filter those out. Finally, <code class="docutils literal notranslate"><span class="pre">sc.pp.filter_genes(adata,</span> <span class="pre">min_cells=10)</span></code> removes genes that are expressed in fewer than 10 cells, helping us focus on genes that are more consistently expressed across our dataset. This step is crucial for ensuring that our downstream analyses are based on reliable data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">max_counts</span><span class="o">=</span><span class="mi">50000</span><span class="p">)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">pct_counts_mt</span> <span class="o">&lt;</span> <span class="mi">20</span><span class="p">]</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/harald/Documents/opt/scllm/.venv/lib/python3.12/site-packages/scanpy/preprocessing/_simple.py:287: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.
  adata.var[&#34;n_cells&#34;] = number
</pre></div></div>
</div>
<p>In this part of the script, we first normalize the total counts in our <code class="docutils literal notranslate"><span class="pre">adata</span></code> object using <code class="docutils literal notranslate"><span class="pre">sc.pp.normalize_total(adata,</span> <span class="pre">inplace=True)</span></code>. This step ensures that the total expression levels across all genes are comparable, which is crucial for downstream analyses. Next, we apply a log transformation with <code class="docutils literal notranslate"><span class="pre">sc.pp.log1p(adata)</span></code>, which helps to stabilize variance and make the data more normally distributed, especially for highly skewed count data.</p>
<p>Then, we identify the highly variable genes using <code class="docutils literal notranslate"><span class="pre">sc.pp.highly_variable_genes(adata,</span> <span class="pre">flavor=&quot;seurat&quot;,</span> <span class="pre">n_top_genes=2000,</span> <span class="pre">subset=True)</span></code>. This function selects the top 2000 genes that show the most variation across cells, which are often the most informative for distinguishing different cell types or states. By subsetting the data to these genes, we focus our analysis on the most relevant features.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">flavor</span><span class="o">=</span><span class="s2">&quot;seurat&quot;</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">2000</span><span class="p">,</span> <span class="n">subset</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>In this part of the code, we’re using the <code class="docutils literal notranslate"><span class="pre">sc.pl.spatial</span></code> function from the Scanpy library to visualize our spatial transcriptomics data. Specifically, we’re plotting the spatial distribution of two metrics: <code class="docutils literal notranslate"><span class="pre">total_counts</span></code> and <code class="docutils literal notranslate"><span class="pre">n_genes_by_counts</span></code>.</p>
<p><code class="docutils literal notranslate"><span class="pre">total_counts</span></code> gives us an idea of the overall expression level in each spot, while <code class="docutils literal notranslate"><span class="pre">n_genes_by_counts</span></code> shows how many different genes are detected in those spots. By visualizing these metrics, we can quickly assess the quality of our data and identify any potential issues, like low-quality spots with few genes detected. This step is crucial for understanding the dataset before diving deeper into analyses like NMF.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">img_key</span><span class="o">=</span><span class="s2">&quot;hires&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;total_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/var/folders/_4/qhxlw4x53134ffft2v5pt3nh0000gn/T/ipykernel_55465/3976244732.py:1: FutureWarning: Use `squidpy.pl.spatial_scatter` instead.
  sc.pl.spatial(adata, img_key=&#34;hires&#34;, color=[&#34;total_counts&#34;, &#34;n_genes_by_counts&#34;])
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_nmf_22_1.png" src="../_images/notebooks_nmf_22_1.png" />
</div>
</div>
</section>
<section id="Applying-NMF">
<h2>Applying NMF<a class="headerlink" href="#Applying-NMF" title="Permalink to this heading">#</a></h2>
<p>In this part of the script, we’re setting up the Non-negative Matrix Factorization (NMF) model. By specifying <code class="docutils literal notranslate"><span class="pre">n_components=10</span></code>, we’re telling the model to identify 10 distinct factors or components from our data, which is crucial for uncovering hidden patterns in the gene expression data from the mouse brain. The <code class="docutils literal notranslate"><span class="pre">max_iter=1000</span></code> parameter allows the algorithm to run for up to 1000 iterations to ensure it converges to a stable solution. After this setup, we fit the NMF model to our data
matrix (<code class="docutils literal notranslate"><span class="pre">adata.X.toarray()</span></code>) and transform it, which gives us the factor scores stored in <code class="docutils literal notranslate"><span class="pre">Z</span></code>. These scores help us understand the underlying structure of the data, which we can then visualize and annotate.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">nmf</span> <span class="o">=</span> <span class="n">NMF</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">max_iter</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>In these lines, we’re applying Non-negative Matrix Factorization (NMF) to our gene expression data stored in <code class="docutils literal notranslate"><span class="pre">adata.X</span></code>. The <code class="docutils literal notranslate"><span class="pre">fit_transform</span></code> method decomposes the data into two matrices: <code class="docutils literal notranslate"><span class="pre">Z</span></code> and <code class="docutils literal notranslate"><span class="pre">W</span></code>. Here, <code class="docutils literal notranslate"><span class="pre">Z</span></code> represents the latent features or components (in our case, 10 components since we set <code class="docutils literal notranslate"><span class="pre">n_components=10</span></code>), which capture the underlying structure of the data. The <code class="docutils literal notranslate"><span class="pre">W</span></code> matrix contains the weights for each gene in relation to these components. Essentially, we’re reducing the
dimensionality of our data while preserving important patterns, which we can later use for visualization and annotation. This step is crucial for understanding the spatial transcriptomics data from the mouse brain.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Z</span> <span class="o">=</span> <span class="n">nmf</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">())</span>
<span class="n">W</span> <span class="o">=</span> <span class="n">nmf</span><span class="o">.</span><span class="n">components_</span>
</pre></div>
</div>
</div>
<p>In these lines, we’re storing the results of our NMF (Non-negative Matrix Factorization) analysis. First, we save the transformed data, <code class="docutils literal notranslate"><span class="pre">Z</span></code>, into <code class="docutils literal notranslate"><span class="pre">adata.obsm['X_nmf']</span></code>, which holds our NMF components for further analysis. Next, we store the NMF components themselves, <code class="docutils literal notranslate"><span class="pre">W</span></code>, in <code class="docutils literal notranslate"><span class="pre">adata.varm['NMF']</span></code>, allowing us to reference the basis vectors later. Finally, we concatenate the NMF results with the existing observation data in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code>, creating new columns named <code class="docutils literal notranslate"><span class="pre">NMF0</span></code> to <code class="docutils literal notranslate"><span class="pre">NMF9</span></code>.
This way, we can easily access and visualize the NMF factors alongside our original data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_nmf&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">Z</span>
<span class="n">adata</span><span class="o">.</span><span class="n">varm</span><span class="p">[</span><span class="s1">&#39;NMF&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">W</span><span class="o">.</span><span class="n">T</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">Z</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;NMF</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">)])],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>In this part of the script, we’re looking at the <code class="docutils literal notranslate"><span class="pre">adata</span></code> object, which contains our processed spatial transcriptomics data from the mouse brain. After performing quality control, normalization, and NMF (Non-negative Matrix Factorization) to reduce dimensionality, we store the NMF results in <code class="docutils literal notranslate"><span class="pre">adata.obsm['X_nmf']</span></code> for the transformed data and <code class="docutils literal notranslate"><span class="pre">adata.varm['NMF']</span></code> for the components. We also add the NMF factors as new columns in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code>, which allows us to visualize and analyze these
factors later. The subsequent lines use the <code class="docutils literal notranslate"><span class="pre">scllm</span></code> package to annotate these NMF factors, helping us understand their biological significance. Finally, we visualize the factor annotations with strip plots and embeddings, giving us insights into the spatial distribution of these factors in the brain tissue.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 2448 × 2000
    obs: &#39;in_tissue&#39;, &#39;array_row&#39;, &#39;array_col&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;, &#39;total_counts_mt&#39;, &#39;log1p_total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;n_counts&#39;, &#39;NMF0&#39;, &#39;NMF1&#39;, &#39;NMF2&#39;, &#39;NMF3&#39;, &#39;NMF4&#39;, &#39;NMF5&#39;, &#39;NMF6&#39;, &#39;NMF7&#39;, &#39;NMF8&#39;, &#39;NMF9&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;mt&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;spatial&#39;, &#39;log1p&#39;, &#39;hvg&#39;
    obsm: &#39;spatial&#39;, &#39;X_nmf&#39;
    varm: &#39;NMF&#39;
</pre></div></div>
</div>
<p>In this part of the script, we’re visualizing the results of our NMF (Non-negative Matrix Factorization) analysis on the spatial transcriptomics data. The <code class="docutils literal notranslate"><span class="pre">sc.pl.spatial</span></code> function is used to create a spatial plot of the mouse brain data, where we specify the <code class="docutils literal notranslate"><span class="pre">img_key=&quot;hires&quot;</span></code> to use the high-resolution image of the tissue. The <code class="docutils literal notranslate"><span class="pre">color</span></code> parameter lists the NMF components we want to visualize—specifically, NMF0 through NMF9. Each of these components represents a different pattern of gene
expression across the spatial locations in the brain. By plotting them, we can see how these patterns are distributed in the tissue, helping us understand the underlying biological processes.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">img_key</span><span class="o">=</span><span class="s2">&quot;hires&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;NMF0&quot;</span><span class="p">,</span> <span class="s2">&quot;NMF1&quot;</span><span class="p">,</span> <span class="s2">&quot;NMF2&quot;</span><span class="p">,</span> <span class="s2">&quot;NMF3&quot;</span><span class="p">,</span> <span class="s2">&quot;NMF4&quot;</span><span class="p">,</span> <span class="s2">&quot;NMF5&quot;</span><span class="p">,</span> <span class="s2">&quot;NMF6&quot;</span><span class="p">,</span> <span class="s2">&quot;NMF7&quot;</span><span class="p">,</span> <span class="s2">&quot;NMF8&quot;</span><span class="p">,</span> <span class="s2">&quot;NMF9&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/var/folders/_4/qhxlw4x53134ffft2v5pt3nh0000gn/T/ipykernel_55465/4223846374.py:1: FutureWarning: Use `squidpy.pl.spatial_scatter` instead.
  sc.pl.spatial(adata, img_key=&#34;hires&#34;, color=[&#34;NMF0&#34;, &#34;NMF1&#34;, &#34;NMF2&#34;, &#34;NMF3&#34;, &#34;NMF4&#34;, &#34;NMF5&#34;, &#34;NMF6&#34;, &#34;NMF7&#34;, &#34;NMF8&#34;, &#34;NMF9&#34;])
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_nmf_33_1.png" src="../_images/notebooks_nmf_33_1.png" />
</div>
</div>
</section>
<section id="Annotating-factors">
<h2>Annotating factors<a class="headerlink" href="#Annotating-factors" title="Permalink to this heading">#</a></h2>
<p>In this part of the code, we use the <code class="docutils literal notranslate"><span class="pre">annotate_factor</span></code> function from the <code class="docutils literal notranslate"><span class="pre">scllm</span></code> package to annotate the factors we obtained from our NMF analysis. We pass our <code class="docutils literal notranslate"><span class="pre">adata</span></code> object, which contains the NMF results, and specify <code class="docutils literal notranslate"><span class="pre">varm_key=&quot;NMF&quot;</span></code> to indicate where the NMF components are stored. The <code class="docutils literal notranslate"><span class="pre">llm=llm</span></code> argument uses our pre-defined language model for generating annotations. The <code class="docutils literal notranslate"><span class="pre">sign=&quot;+&quot;</span></code> parameter indicates that we want to focus on positive contributions of the factors, and
<code class="docutils literal notranslate"><span class="pre">num_samples=10</span></code> tells the function to generate 10 samples for annotation. This step helps us interpret the biological significance of the NMF components by providing meaningful labels, making it easier to understand the underlying patterns in our spatial transcriptomics data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">annotate_factor</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">varm_key</span><span class="o">=</span><span class="s2">&quot;NMF&quot;</span><span class="p">,</span>
    <span class="n">llm</span><span class="o">=</span><span class="n">llm</span><span class="p">,</span>
    <span class="n">sign</span><span class="o">=</span><span class="s2">&quot;+&quot;</span><span class="p">,</span>
    <span class="n">num_samples</span><span class="o">=</span><span class="mi">10</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>In this part of the code, <code class="docutils literal notranslate"><span class="pre">adata.uns['factor_annotation']['mapping']</span></code>, we are accessing the annotations generated by the <code class="docutils literal notranslate"><span class="pre">scllm</span></code> package after we performed NMF (Non-negative Matrix Factorization) on our spatial transcriptomics data. Specifically, this line retrieves the mapping of the NMF factors to biological annotations, which helps us understand what each factor represents in terms of gene expression patterns. This is crucial for interpreting the results of our NMF analysis, as it
connects the numerical factors to meaningful biological insights. Essentially, we’re looking at how the different components we extracted relate to specific biological features in the mouse brain dataset.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;factor_annotation&#39;</span><span class="p">][</span><span class="s1">&#39;mapping&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;0+&#39;: &#39;Erythrocyte&#39;,
 &#39;1+&#39;: &#39;Oligodendrocyte&#39;,
 &#39;2+&#39;: &#39;Medium Spiny Neuron&#39;,
 &#39;3+&#39;: &#39;Purkinje neuron&#39;,
 &#39;4+&#39;: &#39;Excitatory Neuron&#39;,
 &#39;5+&#39;: &#39;Neuroendocrine Cell&#39;,
 &#39;6+&#39;: &#39;Astrocyte&#39;,
 &#39;7+&#39;: &#39;Hepatocyte&#39;,
 &#39;8+&#39;: &#39;Excitatory Neuron&#39;,
 &#39;9+&#39;: &#39;Inhibitory Interneuron&#39;}
</pre></div></div>
</div>
<p>In this part of the script, we use the <code class="docutils literal notranslate"><span class="pre">sl.pl.factor_stripplot</span></code> function from the <code class="docutils literal notranslate"><span class="pre">scllm</span></code> package to visualize the results of our NMF (Non-negative Matrix Factorization) analysis. Specifically, we’re plotting the NMF components stored in <code class="docutils literal notranslate"><span class="pre">X_nmf</span></code> against the annotations generated by the <code class="docutils literal notranslate"><span class="pre">factor_annotation</span></code> step. This helps us see how the different factors identified by NMF relate to the spatial distribution of the data in our mouse brain dataset. The strip plot gives us a clear view of
the expression patterns across the different factors, making it easier to interpret the biological significance of our NMF results. Essentially, we’re visualizing the relationship between the latent factors and the spatial context of the cells in our dataset.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">factor_stripplot</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">obsm_key</span><span class="o">=</span><span class="s2">&quot;X_nmf&quot;</span><span class="p">,</span>
    <span class="n">annotation_key</span><span class="o">=</span><span class="s2">&quot;factor_annotation&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_nmf_40_0.png" src="../_images/notebooks_nmf_40_0.png" />
</div>
</div>
<p>In this part of the script, we use the <code class="docutils literal notranslate"><span class="pre">sl.pl.factor_embedding</span></code> function to visualize the NMF (Non-negative Matrix Factorization) results in a spatial context. We pass our <code class="docutils literal notranslate"><span class="pre">adata</span></code> object, which contains the NMF results stored in <code class="docutils literal notranslate"><span class="pre">obsm_key=&quot;X_nmf&quot;</span></code>, and specify that we want to visualize it based on the spatial coordinates of the data. The <code class="docutils literal notranslate"><span class="pre">annotation_key=&quot;factor_annotation&quot;</span></code> tells the function to use the annotations generated by the <code class="docutils literal notranslate"><span class="pre">scllm</span></code> package to color the plot. This helps us see
how the different NMF factors are distributed across the spatial tissue sections of the mouse brain. Finally, <code class="docutils literal notranslate"><span class="pre">plt.tight_layout()</span></code> ensures that the plot elements are nicely arranged without overlapping.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">sl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">factor_embedding</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">obsm_key</span><span class="o">=</span><span class="s2">&quot;X_nmf&quot;</span><span class="p">,</span>
    <span class="n">basis</span><span class="o">=</span><span class="s1">&#39;spatial&#39;</span><span class="p">,</span>
    <span class="n">annotation_key</span><span class="o">=</span><span class="s2">&quot;factor_annotation&quot;</span>
<span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_nmf_42_0.png" src="../_images/notebooks_nmf_42_0.png" />
</div>
</div>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="pbmc3k.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Annotating the PMBC3K dataset using PCA and scllm</p>
      </div>
    </a>
    <a class="right-next"
       href="angelidis.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Analysing lung cells from aging mice with scllm</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Summary">Summary</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Data-preprocessing">Data preprocessing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Applying-NMF">Applying NMF</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Annotating-factors">Annotating factors</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Harald Vohringer
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2025, Harald Vohringer.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>